### ISSUE TITLE
metrics: merge cost, scanned elements, base case cost

### ISSUE THREAD

**Issue by roarin-roran** [roarin-roran/Sorting#82]
add counters for merge cost, scanned elements, base case cost:

merge cost: number of elements sent to merger

scanned elements: number of elements scanned from _memory_ - repeated access of an element likely to be cached doesn't count.
(@sebawild should this be done with a wrapper like #81 ?)

base case cost: number of elements sent to base case (probably insertion sort).

as merger and base case both depend on more than the number of elements, should we add extra fields here, for example record the lengths of all elements sent to the merger and base case, to calculate average lengths or something like that?

**Comment by sebawild** [comment]
No, scanned elements cannot be handled with a wrapper easily; must be annotated manually.
For merging, it coincides with merge cost, but it should also take base case costs into account.

You *could* instead of scanned elements use a simulated simple cache and use that count cache missed for access to the main array(s); I've done that when we proposed scanned elements as a measure in Quicksort to demonstrate that it can predict cache missed well. I think it is not that useful here.

**Comment by sebawild** [comment]
For base case contributions to scanned elements, you have to consider the specific algorithm. It might be worth counting this separately, i.e., base case costs as you have it (length of input), but also scanned elements in base case. For Insertionsort, it would be roughly quadratic (but should be incremented step by step as you scan the array).

So the overall scanned elements are only computed after the fact and can be taken as either the sum of merge cost and base case cost, or the sum of merge cost and base case scanned elements.

**Comment by roarin-roran** [comment]
sounds good, I think I know roughly what I'm doing with this now.
